scm-webapp/src/main/java/sonia/scm/security/gpg/PublicKeyStore.java

```diff
@@ -0,0 +1,131 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package sonia.scm.security.gpg;
+
+import org.bouncycastle.openpgp.PGPPublicKey;
+import sonia.scm.ContextEntry;
+import sonia.scm.event.ScmEventBus;
+import sonia.scm.repository.Person;
+import sonia.scm.security.NotPublicKeyException;
+import sonia.scm.security.PublicKeyCreatedEvent;
+import sonia.scm.security.PublicKeyDeletedEvent;
+import sonia.scm.store.DataStore;
+import sonia.scm.store.DataStoreFactory;
+import sonia.scm.user.UserPermissions;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static sonia.scm.security.gpg.KeysExtractor.extractPublicKey;
+
+@Singleton
+public class PublicKeyStore {
+
+  private static final String STORE_NAME = "gpg_public_keys";
+  private static final String SUBKEY_STORE_NAME = "gpg_public_sub_keys";
+
+  private final DataStore<RawGpgKey> store;
+  private final DataStore<MasterKeyReference> subKeyStore;
+  private final ScmEventBus eventBus;
+
+  @Inject
+  public PublicKeyStore(DataStoreFactory dataStoreFactory, ScmEventBus eventBus) {
+    this.store = dataStoreFactory.withType(RawGpgKey.class).withName(STORE_NAME).build();
+    this.subKeyStore = dataStoreFactory.withType(MasterKeyReference.class).withName(SUBKEY_STORE_NAME).build();
+    this.eventBus = eventBus;
+  }
+
+  public RawGpgKey add(String displayName, String username, String rawKey) {
+    return add(displayName, username, rawKey, false);
+  }
+
+  public RawGpgKey add(String displayName, String username, String rawKey, boolean readonly) {
+    UserPermissions.changePublicKeys(username).check();
+
+    if (!rawKey.contains("PUBLIC KEY")) {
+      throw new NotPublicKeyException(ContextEntry.ContextBuilder.entity(RawGpgKey.class, displayName).build(), "The provided key is not a public key");
+    }
+
+    Keys keys = Keys.resolve(rawKey);
+    String master = keys.getMaster();
+
+    for (String subKey : keys.getSubs()) {
+      subKeyStore.put(subKey, new MasterKeyReference(master));
+    }
+
+    RawGpgKey key = new RawGpgKey(master, displayName, username, rawKey, getContactsFromPublicKey(rawKey), Instant.now(), readonly);
+
+    store.put(master, key);
+    eventBus.post(new PublicKeyCreatedEvent(new DefaultPublicKey(key.getId(), key.getOwner(), key.getRaw(), key.getContacts())));
```

Comment:
A util method for creating a DefaultPublicKey from a RawGpgKey should make the code more readable.